tests/cpp_validation_tests/test_validation.cpp

```diff
@@ -40,238 +86,263 @@ auto read_json(std::filesystem::path const& path) {
-        for (Idx i{}; i != reference_result.n_components(); ++i) {
-            auto const& component_info = reference_result.get_component_info(i);
-            MetaComponent const& component_meta = *component_info.component;
-            auto const& ref_buffer = reference_result.get_buffer(i);
-            auto const& buffer = result.get_buffer(component_meta.name);
-            Idx const elements_per_scenario = component_info.elements_per_scenario;
-            assert(elements_per_scenario >= 0);
-            // offset scenario
-            RawDataConstPtr const result_ptr =
-                reinterpret_cast<char const*>(buffer.data) + elements_per_scenario * scenario * component_meta.size;
-            RawDataConstPtr const reference_result_ptr =
-                reinterpret_cast<char const*>(ref_buffer.data) + elements_per_scenario * scenario * component_meta.size;
-            // loop all attribute
-            for (MetaAttribute const& attr : component_meta.attributes) {
+
+    DatasetConst const& result = owning_result.const_dataset.value();
+    auto const& result_info = result.get_info();
+    auto const& result_name = result_info.name();
+    Idx const result_batch_size = result_info.batch_size();
+    auto const& storage = owning_result.storage;
+
+    DatasetConst const& reference_result = owning_reference_result.const_dataset.value();
+    auto const& reference_result_info = reference_result.get_info();
+    auto const& reference_result_name = reference_result_info.name();
+    auto const& reference_storage = owning_reference_result.storage;
+    CHECK(storage.buffers.size() == reference_storage.buffers.size());
+
+    // loop through all scenarios
+    for (Idx scenario_idx{}; scenario_idx < result_batch_size; ++scenario_idx) {
+        // loop through all components
+        for (Idx component_idx{}; component_idx < reference_result_info.n_components(); ++component_idx) {
+            auto const& component_name = reference_result_info.component_name(component_idx);
+            auto const* const component_meta = MetaData::get_component_by_name(reference_result_name, component_name);
+
+            auto const& ref_buffer = reference_storage.buffers.at(component_idx);
+            auto const& buffer = storage.buffers.at(component_idx);
+            Idx const elements_per_scenario = reference_result_info.component_elements_per_scenario(component_idx);
+            CHECK(elements_per_scenario >= 0);
+            // loop through all attributes
+            for (Idx attribute_idx{}; attribute_idx < MetaData::n_attributes(component_meta); ++attribute_idx) {
+                auto const* const attribute_meta = MetaData::get_attribute_by_idx(component_meta, attribute_idx);
+                auto attribute_type = MetaData::attribute_ctype(attribute_meta);
+                auto const& attribute_name = MetaData::attribute_name(attribute_meta);
                 // TODO skip u angle, need a way for common angle
```

Comment:
Not from this PR, but got curious: was it skipping that is TODO, or a way for common angle TODO? If latter:
```suggestion
                // TODO need a way for common angle: u angle skipped for now
```

Reply:
the latter